### ISSUE TITLE
Consider using the websocket api for history data

### ISSUE THREAD

**Issue by bdraco** [alexarch21/history-explorer-card#156]
https://github.com/home-assistant/core/blob/9672b5f02cd453f0ea4b1f6a6a32f47098ad87ad/homeassistant/components/history/websocket_api.py#L100

The current call is using an HTTP GET which is much more expensive and slower:
https://github.com/alexarch21/history-explorer-card/blob/2c6965ff1e22db9ded474a4179af27dfb14e0d7f/src/history-explorer-card.js#L1609

**Comment by alexarch21** [comment]
Interesting. I have been using the WS api in a dev build before, but experienced strange results with concurrent calls. That was quite a while ago though, so it might have changed since.

Any reason behind the performance difference ? I would expect both calling the same backend functions. Or is it purely the HTTP REST overhead ?

**Comment by alexarch21** [comment]
Makes sense, thanks for the info. I'll switch to the WS API for normal history calls on the next release. It's cleaner anyway to use the same API for both history and statistics.

**Comment by alexarch21** [comment]
I have done some performance tests on using the WS API for history requests instead of the REST API. I did two test sets, one on requesting the most recent 24 hour period (most likely in the DB cache ?) and one requesting the last 3 weeks (which is the purge days period of my SQLite DB). This was performed on a set of 32 entities requested in a single call.

The DB is stored on a spinning disk NAS, mapped over NFS, on a 1Gbit link. That will obviously add overhead, but I would assume this overhead to be equal for both API calls and thus irrelevant for the comparison.

Only the DB request time was considered (time from issued request to successful callback), card-internal post-processing of the data was not measured. Sample size for the test was 30.

24 Hour period results:
```
  REST:  mean = 0.282 secs, σ = 0.26
  WS:    mean = 0.441 secs, σ = 0.35
```

21 Day period results:
```
  REST: mean = 3.1 secs, σ = 0.23
  WS:   mean = 2.7 secs, σ = 0.24
```

So this is interesting. In fact, the mean of the WS API calls was *slower* on small requests. This is mainly due to large outliers (see the std deviation). While most requests took around the same time as their REST counterparts, some WS requests took almost a second. These outliers skewed the mean. I did not observe these outliers with the REST API over the same requested time period. This is actually visually noticeable when you fast scroll through the history by sliding around when viewing a single day worth of data (which is the default). Since at that range the card typically streams in the data in small 24 blocks, the WS API calls introduces visible brief lags.

When requesting 21 day periods, the WS API is faster, but not by much. The performance increase is not really noticeable in practice. What kind of speedup should I expect here ?

The WS call data is as follows:
```
{
  type: 'history/history_during_period',
  start_time: t0,
  end_time: t1,
  minimal_response: true,
  no_attributes: true,
  entity_ids: l
}
```

**Comment by bdraco** [comment]
How are you measuring the REST and WS socket api calls?

**Comment by alexarch21** [comment]
Client side, Date.now() call before issuing the request and another first thing in the callback. There's some variation to be expected here especially on the small time frames (Date.now() is hardly a performance counter and the browser is not an RT OS), thus the larger sample size. The results were surprisingly consistent though, except for the WS outliers.

**Comment by bdraco** [comment]
How does that compare to what the browser shows for request time in the inspector?

**Comment by alexarch21** [comment]
Good question. Gut feeling is that it won't be far off, as code overhead outside of the request is minimal. I'll check tomorrow.

> Also is HA running on the local machine or is the LAN in the test?

Test is running on LAN. Host is a Pi4 (HA Core in venv), client is my dev PC. Dedicated single connection to the HA backend, the Pi4 is a dedicated dev system on an isolated subnet.

> Also I assume you are testing with no concurrency and all http connection slots are available with the chrome defaults of 6 connections per host?

Correct. 

> Also which version of HA are you testing against?

2023.3.0. Should I try the most recent dot release instead ?

> Are you running a dev or production frontend build?

Production build.

**Comment by bdraco** [comment]
Those results are unexpected for sure.  It would be great to get a py-spy to see more of what's going on to see if there is some type of internal contention on the websocket and that won't only affect history but all of the frontend.

https://community.home-assistant.io/t/instructions-to-install-py-spy-on-haos/480473

Note that 0.3.14 has since been released since the above was published.

The rest api has to do date time conversions from the database timestamp a to python datetime object and than format everything as isoformat time strings.  

The web socket api doesn't have the overhead.

From a processing perspective the code run time to generate the rest response is always going to be longer.

**Comment by alexarch21** [comment]
About the moderate speed increase for the 21 day request, could the reason be the fact that the overhead from accessing the DB file over NFS is so large that it overshadows the API performance differences ? Possibly local storage of the DB (SSD) would yield much larger differences.

**Comment by bdraco** [comment]
> About the moderate speed increase for the 21 day request, could the reason be the fact that the overhead from accessing the DB file over NFS is so large that it overshadows the API performance differences ? Possibly local storage of the DB (SSD) would yield much larger differences.

I was wondering why the times were so high. The I/O likely shadows any processing time.

**Comment by bdraco** [comment]
It would also good to try with 2023.4.x dev since the history backend calls have all be reworked 5 days ago https://github.com/home-assistant/core/blame/dev/homeassistant/components/recorder/history/modern.py

**Comment by alexarch21** [comment]
To be honest, NFS is a terrible choice to map an SQLite file. But it is what it is (and it never really bothered me on my day to day production system). I'll see if I can find a spare SSD somewhere around the office and hook it up to the Pi with a local copy of the DB. That should make speed differences in data processing much more visible.

> It would also good to try with 2023.4.x dev since the history backend calls have all be reworked 5 days ago 

Oh nice ! I'll definitely try that.

I mean I'm going to switch the next release of the card to the WS API either way, simply because of consistency reasons and because I assume this is going to be the way forward in the future for HA frontend access (which would make sense). I'll just try to get to the bottom of the random spikes on small chunks, because those are weird.

**Comment by bdraco** [comment]
There is some code to coalesce the websocket messages if you're getting a burst of them to prevent HA from getting overloaded. 


If you open the inspector and watch the web socket, it should be pretty obvious if that's what's happening

**Comment by alexarch21** [comment]
Ok, noted. I need to do some more tests. This discussion gave me some interesting ideas about things to investigate further. Updates about my findings will follow. Thanks for the assistance !

**Comment by alexarch21** [comment]
Some preliminary results from additional testing by observing the WS activity with inspector and correlating the timing to the JS function timing.

The timing acquired from the JS code above equals almost perfectly the timing from the raw websocket requests in the stream, with a negligible few ms of call overhead. There is also no WS contention or coalescing on this particular messages. The time spent is 99% in the backend for processing the request, the WS is in wait state during almost the entire request - reply cycle. So far, so expected. This also rules out client side issues.

The timing spikes I encountered on the 24h history fetches are also fully on the backend. Identical WS history requests will usually complete within 250-300 ms with the given test data. But from time to time, the same request will take >=800ms in the backend:

306 ms:

![image](https://user-images.githubusercontent.com/60828821/226212325-4e038d26-0144-4e19-bb5b-ac08964d9a6b.png)

799 ms:

![image](https://user-images.githubusercontent.com/60828821/226212337-126796ca-a99e-47ea-9a90-19d2a6d7b2ce.png)

More server side tests would be needed to trace this, but I personally think this is a spike generated by the NFS share on my system, possibly due to FS I/O or a stall due to a concurrent write. In other words, it's not likely to occur with most user systems out there. I do not see this issue as a roadblock switching the card to WS history.

I'm still going to do a quick test with a copy of the test DB on a ramdisk to rule out anything else HA side (and out of curiosity) and also check out the 2023.4 dev build. I'm targeting the new card build for end of the week, so I don't want to spend too much time on what seems an issue mostly tied to my own setup.

**Comment by bdraco** [comment]
Can you see if its better with this patch ?

https://github.com/home-assistant/core/pull/89905

**Comment by alexarch21** [comment]
Sure, let me try this and get back to you.

**Comment by alexarch21** [comment]
Implemented in V1.0.45.

**Comment by bdraco** [comment]
Thanks. Websocket bottleneck should be fixed in 2023.4.x

**Comment by alexarch21** [comment]
Yeah I saw your fix was merged to core. Excellent.
